db/migrations/versions/013_all_event_tables_prog_j_id.py

```diff
@@ -0,0 +1,44 @@
+"""empty message
+
+Revision ID: 013_all_event_tables_prog_j_id
+Revises: 012_event_data_to_jsonb
+Create Date: 2024-02-26 10:29:11.913211
+
+"""
+
+import sqlalchemy as sa
+from alembic import op
+
+import core
+
+# revision identifiers, used by Alembic.
+revision = "013_all_event_tables_prog_j_id"
+down_revision = "012_event_data_to_jsonb"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    with op.batch_alter_table("funding", schema=None) as batch_op:
+        batch_op.add_column(sa.Column("programme_junction_id", core.db.types.GUID(), nullable=True))
+        batch_op.alter_column("project_id", existing_type=sa.UUID(), nullable=True)
+        batch_op.create_index("ix_funding_join_programme_junction", ["programme_junction_id"], unique=False)
+        batch_op.create_foreign_key(
+            batch_op.f("fk_funding_programme_junction_id_programme_junction"),
+            "programme_junction",
+            ["programme_junction_id"],
+            ["id"],
+            ondelete="CASCADE",
+        )
+
+    # ### end Alembic commands ###
+
+
+def downgrade():
+    with op.batch_alter_table("funding", schema=None) as batch_op:
```

Comment:
we're in a weird situation here, whereby this downgrade script will work ok until we add any programme-level data, then it will crash because you won't be able to make the project_id column non-nullable again until you delete out the rows that have no project_id.

It's very unlikely that we would downgrade once we've started ingesting data, so it's probably not an issue. But 1 thing we could do is DROP the rows that would break it before changing it back to non-nullable - It's not ideal as that goes against what a data migration should do (preserve data), so I guess I'm leaning towards leaving it and letting the programmer that reverts the code deal with the implications of dropping those rows (rather than it being automatic)

Reply:
i.e. deal with the ensuing DB error traceback

Reply:
I contend we should leave as-is or else someone in the future, unaware of the minutiae of each migration script, might rollback pre-013 for whatever reason and inadvertently wipe data without knowing about it, and never subsequently re-ingest it. At least this way the error will prompt an understanding of what is happening, and enable future contributors to decide the best course of action.

I'll leave a comment in the migration script to make this clear.